File: src/services/nats_service.rs
use nats::asynk::Connection;
use nats::Options;
use nats::Error;
use tokio::task;

pub async fn connect_to_nats(nats_url: &str) -> Result<Connection, Error> {
    task::spawn_blocking(move || Options::new().connect(nats_url))
        .await
        .map_err(|e| Error::new(e.to_string()))?
}

pub async fn publish_to_nats(nats: &Connection, subject: &str, message: &[u8]) -> Result<(), Error> {
    nats.publish(subject, message).await
}

File: Go/services/nats_service.go
package services

import (
    "github.com/nats-io/nats.go"
    "log"
)

func ConnectToNATS(url string) (*nats.Conn, error) {
    nc, err := nats.Connect(url)
    if err != nil {
        return nil, err
    }

    return nc, nil
}

func PublishToNATS(nc *nats.Conn, subject string, message []byte) error {
    err := nc.Publish(subject, message)
    if err != nil {
        log.Printf("Failed to publish to NATS: %v", err)
        return err
    }

    return nil
}